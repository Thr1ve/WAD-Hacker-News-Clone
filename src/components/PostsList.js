import React, { PropTypes } from 'react';
import { connect } from 'react-redux';

import Post from './Post';

// export unconnected component for testing
export const RawPostsList = ({ isLoading, visibleIds, data }) =>
  isLoading ?
    <div className="container">
      LOADING FEED...
    </div> :
    <div className="container">
      {
        visibleIds.map((id, i) => data[id] ? <Post key={i} post={data[id]} /> :
          <Post key={i} post={{url: "loading", title: "loading"}} />
        )
      }
    </div>;

RawPostsList.propTypes = {
  isLoading: PropTypes.bool,
  data: PropTypes.object,
  visibleIds: PropTypes.array
};

function mapStateToProps(state) {
  return {
    isLoading: state.data.loading,
    data: state.data.cachedItems,
    visibleIds: state.ui.visibleItemIds
  };
}

export default connect(mapStateToProps)(RawPostsList);
