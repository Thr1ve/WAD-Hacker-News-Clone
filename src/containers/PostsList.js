import React from 'react';
import { connect } from 'react-redux';

import { initFeed } from '../actions';
import PostsList from '../components/PostsList';

const PostsListContainer = React.createClass({
  componentWillReceiveProps(nextProps) {
    if (this.props.params.feedName !== nextProps.params.feedName) {
      this.props.initFeed(nextProps.params.feedName);
    }
  },

  componentDidMount() {
      this.props.initFeed(this.props.params.feedName);
  },

  render() {
    return (
      <PostsList
        isLoading={this.props.isLoading}
        data={this.props.data}
        visibleIds={this.props.visibleIds}
      />
    );
  }
})

function mapDispatchToProps(dispatch) {
  return {
    initFeed(feedName) { dispatch(initFeed(feedName)) }
  }
}

function mapStateToProps(state) {
  return {
    isLoading: state.data.loading,
    data: state.data.cachedItems,
    visibleIds: state.ui.visibleItemIds
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(PostsListContainer);
