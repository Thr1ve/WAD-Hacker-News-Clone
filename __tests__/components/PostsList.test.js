import { Map, List } from 'immutable';
import React from 'react';
import renderer from 'react-test-renderer';
import PostsList from '../../src/components/PostsList';

jest.mock('../../src/components/Post', () => 'Post');

const data = Map({})
  .set(12345, {})
  .set(13245, {})
  .set(12435, {})
  .set(12354, undefined);

describe('PostsList:snapshot', () => {
  it('should render correctly', () => {
    const component = renderer.create(
      <PostsList isLoading={false} data={data} visibleIds={List([12345, 12435])}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when isLoading is true', () => {
    const component = renderer.create(
      <PostsList isLoading={true} data={data} visibleIds={List([12345, 12435])}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a "loading" Post when there\'s no data for a given id', () => {
    const component = renderer.create(
      <PostsList isLoading={false} data={data} visibleIds={List([12345, 12435, 12354])}/>
    )
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
