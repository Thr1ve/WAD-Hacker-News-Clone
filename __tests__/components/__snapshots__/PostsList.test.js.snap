exports[`PostsList:snapshot should render a "loading" Post when there's no data for a given id 1`] = `
<div
  className="container">
  <Post
    post={Object {}} />
  <Post
    post={Object {}} />
  <Post
    post={
      Object {
        "title": "loading",
        "url": "loading"
      }
    } />
</div>
`;

exports[`PostsList:snapshot should render correctly 1`] = `
<div
  className="container">
  <Post
    post={Object {}} />
  <Post
    post={Object {}} />
</div>
`;

exports[`PostsList:snapshot should render correctly when isLoading is true 1`] = `
<div
  className="container">
  LOADING FEED...
</div>
`;
